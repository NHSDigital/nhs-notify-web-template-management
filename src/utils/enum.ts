export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}
