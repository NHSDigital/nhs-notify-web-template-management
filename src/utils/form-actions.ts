'use server';

import { getAmplifyBackendClient } from '@utils/amplify-utils';
import { DbOperationError } from '@domain/errors';
import { Template, Draft } from './types';
import { logger } from './logger';

export async function createTemplate(
  template: Draft<Template>
): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.create(template);

  if (errors || !data) {
    logger.error('Failed to create template', errors);
    throw new Error('Failed to create new template');
  }
  return data;
}

export async function saveTemplate(template: Template): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.update(template);

  if (errors) {
    logger.error('Failed to save template', errors);
    throw new Error('Failed to save template data');
  }

  if (!data) {
    throw new DbOperationError({
      message:
        'Template in unknown state. No errors reported but entity returned as falsy',
      operation: 'update',
    });
  }
  return data;
}

export async function getTemplate(
  templateId: string
): Promise<Template | undefined> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.get({
      id: templateId,
    });

  if (errors) {
    logger.error('Failed to get template', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve template for ID ${templateId}`);
    return undefined;
  }

  return data;
}

export async function sendEmail(
  templateId: string,
  templateName: string,
  templateMessage: string,
  templateSubjectLine: string | null
) {
  const res = await getAmplifyBackendClient().queries.sendEmail({
    recipientEmail: 'england.test.cm@nhs.net',
    templateId,
    templateName,
    templateMessage,
    templateSubjectLine,
  });

  if (res.errors) {
    logger.error({
      description: 'Error sending email',
      res,
    });
  }
}

export async function getTemplates(): Promise<Template[] | []> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.list();

  if (errors) {
    logger.error('Failed to get templates', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve templates`);
    return [];
  }

  return data;
}
