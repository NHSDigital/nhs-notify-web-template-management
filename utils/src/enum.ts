export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
}

export enum TemplateStatus {
  NOT_YET_SUBMITTED = 'NOT_YET_SUBMITTED',
  SUBMITTED = 'SUBMITTED',
}

export const templateTypeDisplayMappings = (type: TemplateType) =>
  ({
    [TemplateType.NHS_APP]: 'NHS App',
    [TemplateType.SMS]: 'Text message (SMS)',
    [TemplateType.EMAIL]: 'Email',
  })[type];

export const templateStatustoDisplayMappings = (status: TemplateStatus) =>
  ({
    [TemplateStatus.NOT_YET_SUBMITTED]: 'Not yet submitted',
    [TemplateStatus.SUBMITTED]: 'Submitted',
  })[status];
