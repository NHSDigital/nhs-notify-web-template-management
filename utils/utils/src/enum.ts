import {
  TemplateType,
  TemplateStatus,
  LetterType,
  Language,
} from 'nhs-notify-backend-client';

// eslint-disable-next-line unicorn/prefer-export-from
export { TemplateType, TemplateStatus };

export const languageMapping = (language: Language) =>
  ({
    [Language.AR]: 'Arabic',
    [Language.BG]: 'Bulgarian',
    [Language.BN]: 'Bengali',
    [Language.DE]: 'German',
    [Language.EL]: 'Greek',
    [Language.EN]: 'English',
    [Language.ES]: 'Spanish',
    [Language.FA]: 'Persian',
    [Language.FR]: 'French',
    [Language.GU]: 'Gujurati',
    [Language.HI]: 'Hindi',
    [Language.HU]: 'Hungarian',
    [Language.IT]: 'Italian',
    [Language.KU]: 'Kurdish',
    [Language.LT]: 'Lithuanian',
    [Language.LV]: 'Latvian',
    [Language.NE]: 'Nepali',
    [Language.PA]: 'Punjabi',
    [Language.PL]: 'Polish',
    [Language.PT]: 'Portuguese',
    [Language.RO]: 'Romanian',
    [Language.RU]: 'Russian',
    [Language.SK]: 'Slovak',
    [Language.SO]: 'Somali',
    [Language.SQ]: 'Albanian',
    [Language.TA]: 'Tamil',
    [Language.TR]: 'Turkish',
    [Language.UR]: 'Urdu',
    [Language.ZH]: 'Chinese',
  })[language];
const letterTypeMapping = (letterType: LetterType) =>
  ({
    [LetterType.X3]: 'Audio CD',
    [LetterType.Q1]: 'Braille',
    [LetterType.Q4]: 'British Sign Language',
    [LetterType.X0]: 'Standard',
    [LetterType.X1]: 'Large print',
  })[letterType];

export const letterTypeDisplayMappings = (
  letterType: LetterType,
  language: Language
) =>
  language === Language.EN
    ? `${letterTypeMapping(letterType)} letter`
    : `Letter - ${languageMapping(language)}`;

export const templateTypeDisplayMappings = (type: TemplateType) =>
  ({
    NHS_APP: 'NHS App message',
    SMS: 'Text message (SMS)',
    EMAIL: 'Email',
    LETTER: 'Letter',
  })[type];

export const templateStatusToDisplayMappings = (status: TemplateStatus) =>
  ({
    [TemplateStatus.NOT_YET_SUBMITTED]: 'Not yet submitted',
    [TemplateStatus.SUBMITTED]: 'Submitted',
    [TemplateStatus.DELETED]: '', // will not be shown in the UI
  })[status];

export const templateTypeToUrlTextMappings = (type: TemplateType) =>
  ({
    NHS_APP: 'nhs-app',
    SMS: 'text-message',
    EMAIL: 'email',
    LETTER: 'letter',
  })[type];

export const previewTemplatePages = (type: TemplateType) =>
  `preview-${templateTypeToUrlTextMappings(type)}-template`;
export const viewSubmittedTemplatePages = (type: TemplateType) =>
  `view-submitted-${templateTypeToUrlTextMappings(type)}-template`;
