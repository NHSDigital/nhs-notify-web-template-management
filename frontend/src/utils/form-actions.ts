'use server';

import { getSessionServer } from '@utils/amplify-utils';
import {
  CreateUpdateTemplate,
  isTemplateDtoValid,
  RoutingConfig,
  RoutingConfigStatusActive,
  routingConfigurationApiClient,
  TemplateDto,
  ValidatedTemplateDto,
} from 'nhs-notify-backend-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';
import { templateApiClient } from 'nhs-notify-backend-client/src/template-api-client';
import { sortAscByCreatedAt } from './sort';

export async function createTemplate(
  template: CreateUpdateTemplate
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.createTemplate(
    template,
    accessToken
  );

  if (error) {
    logger.error('Failed to create template', { error });
    throw new Error('Failed to create new template');
  }

  return data;
}

export async function uploadLetterTemplate(
  template: CreateUpdateTemplate,
  pdf: File,
  csv: File
) {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.uploadLetterTemplate(
    template,
    accessToken,
    pdf,
    csv?.size > 0 ? csv : undefined
  );

  if (error) {
    logger.error('Failed to create letter template', { error });
    throw new Error('Failed to create new letter template');
  }

  return data;
}

export async function saveTemplate(
  template: Extract<TemplateDto, { templateType: 'EMAIL' | 'SMS' | 'NHS_APP' }>
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.updateTemplate(
    template.id,
    template,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }

  return data;
}

export async function setTemplateToSubmitted(
  templateId: string
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.submitTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }

  return data;
}

export async function setTemplateToDeleted(templateId: string): Promise<void> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { error } = await templateApiClient.deleteTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to save template', { error });
    throw new Error('Failed to save template data');
  }
}

export async function requestTemplateProof(
  templateId: string
): Promise<TemplateDto> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.requestProof(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to request proof', { error });
    throw new Error('Failed to request proof');
  }

  return data;
}

export async function getTemplate(
  templateId: string
): Promise<TemplateDto | undefined> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.getTemplate(
    templateId,
    accessToken
  );

  if (error) {
    logger.error('Failed to get template', { error });
  }

  return data;
}

export async function getTemplates(): Promise<TemplateDto[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await templateApiClient.listTemplates(accessToken);

  if (error) {
    logger.error('Failed to get templates', { error });
    return [];
  }

  const sortedData = data
    .map((template) => isTemplateDtoValid(template))
    .filter(
      (template): template is ValidatedTemplateDto => template !== undefined
    );

  return sortAscByCreatedAt(sortedData);
}

export async function getRoutingConfigs(): Promise<RoutingConfig[]> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.list(accessToken);

  if (error) {
    logger.error('Failed to get routing configuration', {
      error: error,
    });
    return [];
  }

  return sortAscByCreatedAt(data);
}

export async function countRoutingConfigs(
  status: RoutingConfigStatusActive
): Promise<number> {
  const { accessToken } = await getSessionServer();

  if (!accessToken) {
    throw new Error('Failed to get access token');
  }

  const { data, error } = await routingConfigurationApiClient.count(
    accessToken,
    status
  );

  if (error) {
    logger.error(`Failed to count routing configuration for ${status}`, {
      error,
    });
    return 0;
  }

  return data.count;
}
