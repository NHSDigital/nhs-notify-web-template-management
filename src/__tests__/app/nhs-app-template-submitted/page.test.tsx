import nav from 'next/navigation';
import NhsAppTemplateSubmittedPage from '@app/nhs-app-template-submitted/[templateId]/page';
import { render } from '@testing-library/react';
import { getTemplate } from '@utils/form-actions';
import { TemplateStatus, TemplateType } from '@utils/enum';

jest.mock('@utils/form-actions', () => ({
  getTemplate: jest.fn().mockImplementation((templateId: string) => {
    if (templateId === 'template-id') {
      return {
        id: 'template-id',
        version: 1,
        templateType: TemplateType.NHS_APP,
        templateStatus: TemplateStatus.SUBMITTED,
        name: 'template-name',
        message: 'template-message',
      };
    }
  }),
}));

jest.mock('next/navigation', () => ({
  redirect: () => {
    throw new Error('Simulated redirect');
  },

  RedirectType: {
    push: 'push',
    replace: 'replace',
  },
}));

jest.mock('@utils/logger');

test('NhsAppTemplateSubmittedPage', async () => {
  const page = await NhsAppTemplateSubmittedPage({
    params: {
      templateId: 'template-id',
    },
  });

  const container = render(page);

  expect(jest.mocked(getTemplate)).toHaveBeenCalledWith('template-id');
  expect(container.asFragment()).toMatchSnapshot();
});

test('NhsAppTemplateSubmittedPage - should handle invalid template', async () => {
  const redirectSpy = jest.spyOn(nav, 'redirect');

  await expect(
    NhsAppTemplateSubmittedPage({
      params: {
        templateId: 'invalid-template',
      },
    })
  ).rejects.toThrow('Simulated redirect');

  expect(jest.mocked(getTemplate)).toHaveBeenCalledWith('invalid-template');
  expect(redirectSpy).toHaveBeenCalledWith('/invalid-template', 'replace');
});
