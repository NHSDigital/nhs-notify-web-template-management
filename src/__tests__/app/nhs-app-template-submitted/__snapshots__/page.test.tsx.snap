// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NhsAppTemplateSubmittedPage 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <div
        class="notify-confirmation-panel"
      >
        <h1
          class="nhsuk-heading-l nhsuk-u-margin-bottom-0"
          id="template-submitted"
        >
          Template submitted
        </h1>
      </div>
      <h2
        class="nhsuk-heading-xs nhsuk-u-margin-bottom-1"
      >
        Template name
      </h2>
      <p
        id="template-name"
      >
        template-name
      </p>
      <h2
        class="nhsuk-heading-xs nhsuk-u-margin-bottom-1"
      >
        Template ID
      </h2>
      <p
        id="template-id"
      >
        template-id
      </p>
      <a
        href="/create-template"
        id="create-another-template"
      >
        Create another template
      </a>
      <h2
        class="nhsuk-u-margin-top-5"
      >
        What you need to do next
      </h2>
      <p>
        You'll receive a confirmation email, which contains the template name and ID.
      </p>
      <h3>
        If you're currently onboarding
      </h3>
      <p>
        Tell your onboarding manager once you've submitted all your templates.
      </p>
      <h3>
        If you've already onboarded
      </h3>
      <p>
        <a
          href="https://nhsdigitallive.service-now.com/nhs_digital?id=sc_cat_item&sys_id=6208dbce1be759102eee65b9bd4bcbf5"
          id="servicenow-link"
        >
          Raise a request with the service desk
        </a>
         once you've submitted all your templates.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
